src/libsyntax/ext/tt/macro_check.rs

```diff
@@ -0,0 +1,621 @@
+//! Checks that meta-variables in macro definition are correctly declared and used.
+//!
+//! # What is checked
+//!
+//! ## Meta-variables must not be bound twice
+//!
+//! ```
+//! macro_rules! foo { ($x:tt $x:tt) => { $x }; }
+//! ```
+//!
+//! This check is also done at macro instantiation.
+//!
+//! ## Meta-variables must not be free
+//!
+//! ```
+//! macro_rules! foo { () => { $x }; }
+//! ```
+//!
+//! This check is also done at macro instantiation but only if the branch is taken.
+//!
+//! ## Meta-variables must repeat at least as many times as their binder
+//!
+//! ```
+//! macro_rules! foo { ($($x:tt)*) => { $x }; }
+//! ```
+//!
+//! This check is also done at macro instantiation but only if the branch is taken.
+//!
+//! ## Meta-variables must repeat with the same Kleene operators as their binder
+//!
+//! ```
+//! macro_rules! foo { ($($x:tt)+) => { $($x)* }; }
```

Comment:
This seems more like a language problem.
Repetition is determined entirely by the left side, and right side could use some single entirely unrelated symbol to denote repetition
```rust
macro_rules! foo { ($($x:tt)+) => { $($x)~ }; }
```
or no symbol at all
```rust
macro_rules! foo { ($($x:tt)+) => { $($x) }; }
```
I think we should seriously consider the second alternative, macro definitions are notoriously noisy and it would help to remove some of that noise if it is feasible.

Reply:
Right makes sense. Maybe we should discuss this in #61053. Then I'll apply the decision in this PR.

Reply:
This is a future work / RFC material, certainly not for this PR.

Reply:
I see, I removed the lint for this check, but used the information to give a possibly helpful error message in case a meta-variable is not repeating enough. If we don't want that, then we can make the code much simpler by only storing the depth instead of the stack. What do you think?

In particular, could you advise on the following points?

- When a meta-variable is not repeating enough, do we want to show the first missing repetition in the LHS?
    ```
    LL |     ($( $i:ident = $($j:ident),+ );*) => { $( $i = $j; )* };
       |                                -                   ^^
       |                                |
       |                                expected repetition
    ```
- When a meta-variable is not repeating enough, do we want to show the first differing Kleene operator?
    ```
    LL |     ($( $i:ident = $($j:ident),+ );*) => { $( $i = $j; )+ };
       |                                    -               ^^   - first conflicting repetition
       |                                    |
       |                                    first expected repetition
    ```

Not doing the second point permits to only store depth for occurrences (instead of stack). Not doing both points permits in addition to only store the depth for the binders too and also to not have to store the span of the Kleene operators at all. This PR would be much simpler.

Reply:
I would actually prefer to keep the lint for now until we decide what to do. Having the same operators on each side makes macros significantly easier to read and write, IMHO.

Reply:
Let me know if I should restore this check.

Reply:
Yes, please, restore the check, I'll rerun bors try after that.